File: client-py/queen/http/http_client.py
"""
HTTP client with retry, load balancing, and failover support
"""

import asyncio
from typing import Any, Dict, Optional, Union
import httpx

from ..utils import logger
from .load_balancer import LoadBalancer


class HttpClient:
    """HTTP client with retry, load balancing, and failover support"""

    def __init__(
        self,
        *,
        base_url: Optional[str] = None,
        load_balancer: Optional[LoadBalancer] = None,
        timeout_millis: int = 30000,
        retry_attempts: int = 3,
        retry_delay_millis: int = 1000,
        enable_failover: bool = True,
        bearer_token: Optional[str] = None,
    ):
        """
        Initialize HTTP client

        Args:
            base_url: Base URL for single server
            load_balancer: LoadBalancer instance for multiple servers
            timeout_millis: Request timeout in milliseconds
            retry_attempts: Number of retry attempts
            retry_delay_millis: Initial retry delay (exponential backoff)
            enable_failover: Enable automatic failover
            bearer_token: Bearer token for proxy authentication
        """
        self._base_url = base_url
        self._load_balancer = load_balancer
        self._timeout_millis = timeout_millis
        self._retry_attempts = retry_attempts
        self._retry_delay_millis = retry_delay_millis
        self._enable_failover = enable_failover
        self._bearer_token = bearer_token

        # Build headers with optional auth
        headers = {}
        if bearer_token:
            headers["Authorization"] = f"Bearer {bearer_token}"

        # Create httpx.AsyncClient (persistent connection pool)
        self._client = httpx.AsyncClient(
            timeout=httpx.Timeout(timeout_millis / 1000.0),
            limits=httpx.Limits(max_keepalive_connections=10, max_connections=100),
            headers=headers,
        )

        logger.log(
            "HttpClient.constructor",
            {
                "has_load_balancer": load_balancer is not None,
                "base_url": base_url or "load-balanced",
                "timeout_millis": timeout_millis,
                "retry_attempts": retry_attempts,
                "enable_failover": enable_failover,
                "has_auth": bearer_token is not None,
            },
        )

    async def get(
        self,
        path: str,
        request_timeout_millis: Optional[int] = None,
        affinity_key: Optional[str] = None,
    ) -> Any:
        """GET request"""
        return await self._request_with_failover(
            "GET", path, None, request_timeout_millis, affinity_key
        )

    async def post(
        self,
        path: str,
        body: Optional[Dict[str, Any]] = None,
        request_timeout_millis: Optional[int] = None,
        affinity_key: Optional[str] = None,
    ) -> Any:
        """POST request"""
        return await self._request_with_failover(
            "POST", path, body, request_timeout_millis, affinity_key
        )

    async def put(
        self,
        path: str,
        body: Optional[Dict[str, Any]] = None,
        request_timeout_millis: Optional[int] = None,
        affinity_key: Optional[str] = None,
    ) -> Any:
        """PUT request"""
        return await self._request_with_failover(
            "PUT", path, body, request_timeout_millis, affinity_key
        )

    async def delete(
        self,
        path: str,
        request_timeout_millis: Optional[int] = None,
        affinity_key: Optional[str] = None,
    ) -> Any:
        """DELETE request"""
        return await self._request_with_failover(
            "DELETE", path, None, request_timeout_millis, affinity_key
        )

    async def _execute_request(
        self,
        url: str,
        method: str,
        body: Optional[Dict[str, Any]],
        request_timeout_millis: Optional[int],
    ) -> Any:
        """Execute single HTTP request"""
        effective_timeout = request_timeout_millis or self._timeout_millis
        logger.log(
            "HttpClient.request",
            {"method": method, "url": url, "has_body": body is not None, "timeout": effective_timeout},
        )

        try:
            # Set timeout for this specific request
            timeout = httpx.Timeout(effective_timeout / 1000.0)

            # Prepare request
            kwargs: Dict[str, Any] = {
                "method": method,
                "url": url,
                "timeout": timeout,
            }

            if body:
                kwargs["json"] = body

            response = await self._client.request(**kwargs)

            logger.log("HttpClient.response", {"method": method, "url": url, "status": response.status_code})

            # Handle 204 No Content
            if response.status_code == 204:
                return None

            # Handle errors
            if not response.is_success:
                error_msg = f"HTTP {response.status_code}: {response.reason_phrase}"
                try:
                    text = response.text
                    if text:
                        body_data = response.json()
                        error_msg = body_data.get("error", error_msg)
                except Exception:
                    pass

                logger.error(
                    "HttpClient.request",
                    {"method": method, "url": url, "status": response.status_code, "error": error_msg},
                )
                error = httpx.HTTPStatusError(error_msg, request=response.request, response=response)
                raise error

            # Parse successful response
            content_type = response.headers.get("content-type", "")
            content_length = response.headers.get("content-length", "")

            if (
                not content_type
                or "application/json" not in content_type
                or content_length == "0"
            ):
                text = response.text
                if not text:
                    return None
                try:
                    return response.json()
                except Exception:
                    return None

            return response.json()

        except httpx.TimeoutException as e:
            logger.error(
                "HttpClient.request",
                {"method": method, "url": url, "error": "timeout", "timeout": effective_timeout},
            )
            raise
        except Exception as e:
            logger.error("HttpClient.request", {"method": method, "url": url, "error": str(e)})
            raise

    async def _request_with_retry(
        self,
        method: str,
        path: str,
        body: Optional[Dict[str, Any]],
        request_timeout_millis: Optional[int],
    ) -> Any:
        """Execute request with retry logic"""
        last_error: Optional[Exception] = None

        for attempt in range(self._retry_attempts):
            try:
                url = self._get_url() + path
                return await self._execute_request(url, method, body, request_timeout_millis)
            except Exception as error:
                last_error = error

                # Don't retry on client errors (4xx)
                if isinstance(error, httpx.HTTPStatusError):
                    status = error.response.status_code
                    if 400 <= status < 500:
                        raise error

                # Wait before retry (except on last attempt)
                if attempt < self._retry_attempts - 1:
                    delay = (self._retry_delay_millis / 1000.0) * (2**attempt)
                    logger.warn(
                        "HttpClient.retry",
                        {
                            "method": method,
                            "path": path,
                            "attempt": attempt + 1,
                            "delay": delay,
                            "error": str(error),
                        },
                    )
                    await asyncio.sleep(delay)

        logger.error(
            "HttpClient.retry",
            {"method": method, "path": path, "error": "Max retries exceeded", "attempts": self._retry_attempts},
        )
        if last_error:
            raise last_error
        raise Exception("Max retries exceeded")

    async def _request_with_failover(
        self,
        method: str,
        path: str,
        body: Optional[Dict[str, Any]],
        request_timeout_millis: Optional[int],
        affinity_key: Optional[str],
    ) -> Any:
        """Execute request with failover logic"""
        if not self._load_balancer or not self._enable_failover:
            return await self._request_with_retry(method, path, body, request_timeout_millis)

        urls = self._load_balancer.get_all_urls()
        attempted_urls = set()
        last_error: Optional[Exception] = None

        logger.log(
            "HttpClient.failover",
            {"method": method, "path": path, "total_servers": len(urls), "affinity_key": affinity_key},
        )

        for _ in range(len(urls)):
            # Pass affinity key to load balancer for consistent routing
            url = self._load_balancer.get_next_url(affinity_key)

            if url in attempted_urls:
                continue

            attempted_urls.add(url)

            try:
                result = await self._execute_request(url + path, method, body, request_timeout_millis)

                # Mark backend as healthy on success
                self._load_balancer.mark_healthy(url)

                return result
            except Exception as error:
                last_error = error

                # Mark backend as unhealthy on failure (5xx or network errors)
                if isinstance(error, httpx.HTTPStatusError):
                    status = error.response.status_code
                    if status >= 500:
                        self._load_balancer.mark_unhealthy(url)
                elif isinstance(error, (httpx.NetworkError, httpx.TimeoutException)):
                    self._load_balancer.mark_unhealthy(url)

                logger.warn(
                    "HttpClient.failover", {"url": url, "method": method, "path": path, "error": str(error)}
                )
                print(f"Request failed for {url}: {method} {path} - {error}")

                # Don't retry on client errors (4xx)
                if isinstance(error, httpx.HTTPStatusError):
                    status = error.response.status_code
                    if 400 <= status < 500:
                        raise error

                # Continue to next server for server errors or network issues

        logger.error(
            "HttpClient.failover",
            {"method": method, "path": path, "error": "All servers failed", "attempted": len(attempted_urls)},
        )
        if last_error:
            raise last_error
        raise Exception("All servers failed")

    def _get_url(self) -> str:
        """Get URL from load balancer or base URL"""
        if self._load_balancer:
            return self._load_balancer.get_next_url()
        return self._base_url or ""

    def get_load_balancer(self) -> Optional[LoadBalancer]:
        """Get load balancer instance"""
        return self._load_balancer

    async def close(self) -> None:
        """Close HTTP client and connection pool"""
        await self._client.aclose()

File: client-js/client-v2/http/HttpClient.js
/**
 * HTTP client with retry, load balancing, and failover support
 */

import * as logger from '../utils/logger.js'

export class HttpClient {
  #baseUrl
  #loadBalancer
  #timeoutMillis
  #retryAttempts
  #retryDelayMillis
  #enableFailover
  #bearerToken
  #headers

  constructor(options = {}) {
    const {
      baseUrl = null,
      loadBalancer = null,
      timeoutMillis = 30000,
      retryAttempts = 3,
      retryDelayMillis = 1000,
      enableFailover = true,
      bearerToken = null,
      headers = {}
    } = options

    this.#baseUrl = baseUrl
    this.#loadBalancer = loadBalancer
    this.#timeoutMillis = timeoutMillis
    this.#retryAttempts = retryAttempts
    this.#retryDelayMillis = retryDelayMillis
    this.#enableFailover = enableFailover
    this.#bearerToken = bearerToken
    this.#headers = headers || {}
    
    logger.log('HttpClient.constructor', { 
      hasLoadBalancer: !!loadBalancer, 
      baseUrl: baseUrl || 'load-balanced', 
      timeoutMillis, 
      retryAttempts,
      enableFailover,
      hasAuth: !!bearerToken,
      customHeaders: Object.keys(this.#headers).length
    })
  }

  async #executeRequest(url, method, body = null, requestTimeoutMillis = null) {
    const effectiveTimeout = requestTimeoutMillis || this.#timeoutMillis
    logger.log('HttpClient.request', { method, url, hasBody: !!body, timeout: effectiveTimeout })
    
    const controller = new AbortController()
    const timeoutId = setTimeout(() => controller.abort(), effectiveTimeout)

    try {
      const headers = { 'Content-Type': 'application/json' }
      if (this.#bearerToken) {
        headers['Authorization'] = `Bearer ${this.#bearerToken}`
      }
      Object.assign(headers, this.#headers)

      const options = {
        method,
        signal: controller.signal,
        headers
      }

      if (body) {
        options.body = JSON.stringify(body)
      }

      const response = await fetch(url, options)
      
      logger.log('HttpClient.response', { method, url, status: response.status })

      // Handle 204 No Content
      if (response.status === 204) {
        return null
      }

      // Handle errors
      if (!response.ok) {
        const error = new Error(`HTTP ${response.status}: ${response.statusText}`)
        error.status = response.status

        try {
          const text = await response.text()
          if (text) {
            const body = JSON.parse(text)
            error.message = body.error || error.message
          }
        } catch (e) {
          // Ignore JSON parse errors
        }

        logger.error('HttpClient.request', { method, url, status: error.status, error: error.message })
        throw error
      }

      // Parse successful response
      const contentType = response.headers.get('content-type')
      const contentLength = response.headers.get('content-length')

      if (!contentType || !contentType.includes('application/json') || contentLength === '0') {
        const text = await response.text()
        if (!text || text.length === 0) {
          return null
        }
        try {
          return JSON.parse(text)
        } catch (e) {
          return null
        }
      }

      return response.json()

    } catch (error) {
      if (error.name === 'AbortError') {
        const timeoutError = new Error(`Request timeout after ${effectiveTimeout}ms`)
        timeoutError.name = 'AbortError'
        timeoutError.timeout = effectiveTimeout
        logger.error('HttpClient.request', { method, url, error: 'timeout', timeout: effectiveTimeout })
        throw timeoutError
      }
      logger.error('HttpClient.request', { method, url, error: error.message })
      throw error
    } finally {
      clearTimeout(timeoutId)
    }
  }

  async #requestWithRetry(method, path, body = null, requestTimeoutMillis = null) {
    let lastError = null

    for (let attempt = 0; attempt < this.#retryAttempts; attempt++) {
      try {
        const url = this.#getUrl() + path
        return await this.#executeRequest(url, method, body, requestTimeoutMillis)
      } catch (error) {
        lastError = error

        // Don't retry on client errors (4xx)
        if (error.status && error.status >= 400 && error.status < 500) {
          throw error
        }

        // Wait before retry (except on last attempt)
        if (attempt < this.#retryAttempts - 1) {
          const delay = this.#retryDelayMillis * Math.pow(2, attempt)
          logger.warn('HttpClient.retry', { method, path, attempt: attempt + 1, delay, error: error.message })
          await new Promise(resolve => setTimeout(resolve, delay))
        }
      }
    }

    logger.error('HttpClient.retry', { method, path, error: 'Max retries exceeded', attempts: this.#retryAttempts })
    throw lastError
  }

  async #requestWithFailover(method, path, body = null, requestTimeoutMillis = null, affinityKey = null) {
    if (!this.#loadBalancer || !this.#enableFailover) {
      return this.#requestWithRetry(method, path, body, requestTimeoutMillis)
    }

    const urls = this.#loadBalancer.getAllUrls()
    const attemptedUrls = new Set()
    let lastError = null

    logger.log('HttpClient.failover', { method, path, totalServers: urls.length, affinityKey })

    for (let i = 0; i < urls.length; i++) {
      // Pass affinity key to load balancer for consistent routing
      const url = this.#loadBalancer.getNextUrl(affinityKey)

      if (attemptedUrls.has(url)) {
        continue
      }

      attemptedUrls.add(url)

      try {
        const result = await this.#executeRequest(url + path, method, body, requestTimeoutMillis)
        
        // Mark backend as healthy on success
        this.#loadBalancer.markHealthy(url)
        
        return result
      } catch (error) {
        lastError = error
        
        // Mark backend as unhealthy on failure (5xx or network errors)
        if (!error.status || error.status >= 500) {
          this.#loadBalancer.markUnhealthy(url)
        }
        
        logger.warn('HttpClient.failover', { url, method, path, error: error.message })
        console.warn(`Request failed for ${url}: ${method} ${path} - ${error.message}`)

        // Don't retry on client errors (4xx)
        if (error.status && error.status >= 400 && error.status < 500) {
          throw error
        }

        // Continue to next server for server errors or network issues
      }
    }

    logger.error('HttpClient.failover', { method, path, error: 'All servers failed', attempted: attemptedUrls.size })
    throw lastError || new Error('All servers failed')
  }

  #getUrl() {
    if (this.#loadBalancer) {
      return this.#loadBalancer.getNextUrl()
    }
    return this.#baseUrl
  }

  async get(path, requestTimeoutMillis = null, affinityKey = null) {
    return this.#requestWithFailover('GET', path, null, requestTimeoutMillis, affinityKey)
  }

  async post(path, body = null, requestTimeoutMillis = null, affinityKey = null) {
    return this.#requestWithFailover('POST', path, body, requestTimeoutMillis, affinityKey)
  }

  async put(path, body = null, requestTimeoutMillis = null, affinityKey = null) {
    return this.#requestWithFailover('PUT', path, body, requestTimeoutMillis, affinityKey)
  }

  async delete(path, requestTimeoutMillis = null, affinityKey = null) {
    return this.#requestWithFailover('DELETE', path, null, requestTimeoutMillis, affinityKey)
  }

  getLoadBalancer() {
    return this.#loadBalancer
  }
}

